pkg/scheduler/framework/framework.go

```diff
@@ -759,3 +792,129 @@ func (f *framework) downscale(ctx context.Context, scheduled, bound []*fleetv1be
 
 	return nil, sorted[left:], f.markAsUnscheduledFor(ctx, bindingsToDelete)
 }
+
+// runAllPluginsForPickNPlacementType runs all plugins for a scheduling policy of the PickN placement type.
+//
+// Note that all stages are required to run for this placement type.
+func (f *framework) runAllPluginsForPickNPlacementType(
+	ctx context.Context,
+	state *CycleState,
+	policy *fleetv1beta1.ClusterSchedulingPolicySnapshot,
+	numOfClusters int,
+	numOfBoundOrScheduledBindings int,
+	clusters []fleetv1beta1.MemberCluster,
+) (scored ScoredClusters, filtered []*filteredClusterWithStatus, err error) {
+	policyRef := klog.KObj(policy)
+
+	// Calculate the batch size.
+	//
+	// Note that obsolete bindings are not counted.
+	state.desiredBatchSize = numOfClusters - numOfBoundOrScheduledBindings
+
+	// An earlier check guarantees that the desired batch size is always positive; however, the scheduler still
+	// performs a sanity check here; normally this branch will never run.
+	if state.desiredBatchSize <= 0 {
+		err := fmt.Errorf("desired batch size is below zero: %d", state.desiredBatchSize)
+		klog.ErrorS(err, "Failed to calculate desired batch size", "clusterSchedulingPolicySnapshot", policyRef)
+		return nil, nil, controller.NewUnexpectedBehaviorError(err)
+	}
+
+	// Run pre-batch plugins.
+	//
+	// These plugins each yields a batch size limit; the minimum of these limits is used as the actual batch size for
+	// this scheduling cycle.
+	//
+	// Note that any failure would lead to the cancellation of the scheduling cycle.
+	batchSizeLimit, status := f.runPostBatchPlugins(ctx, state, policy)
+	if status.IsInteralError() {
+		klog.ErrorS(status.AsError(), "Failed to run post batch plugins", "clusterSchedulingPolicySnapshot", policyRef)
+		return nil, nil, controller.NewUnexpectedBehaviorError(status.AsError())
+	}
+
+	// A sanity check; normally this branch will never run, as runPostBatchPlugins guarantees that
+	// the batch size limit is never greater than the desired batch size.
+	if batchSizeLimit > state.desiredBatchSize || batchSizeLimit < 0 {
+		err := fmt.Errorf("batch size limit is not valid: %d, desired batch size: %d", batchSizeLimit, state.desiredBatchSize)
+		klog.ErrorS(err, "Failed to set batch size limit", "clusterSchedulingPolicySnapshot", policyRef)
+		return nil, nil, controller.NewUnexpectedBehaviorError(err)
+	}
+	state.batchSizeLimit = batchSizeLimit
```

Comment:
when the batchSizeLimit will be used?

Reply:
It is used when actually picking the clusters.